### ISSUE TITLE
How to pass a switch python argument into input_signature when using tf.function? 

### ISSUE THREAD

**Issue by LionQYW** [keras-team/keras#15751]
Hi, I’m using Tensorflow 2.7.0 currently. There is no any other configuration things related to my issue.

I’m trying to set a switch for a tf.function, like set the input_signature for the call method of a Karas Model, so I can use the tf.saved_model.save API to save it. But it seems like I can not pass a python argument into the input_signature.

For example, I built a Transformer translation Model, I set an argument named “istraining” for the call method of this Model. When istraining = True, the Model run train mode and use the “teacher forcing” pattern to decode target language input. When istraining = False, the model run translating mode and use the “auto regression” pattern to generate target language.

But I can not find a way to pass the python True/False argument into the input_signature of the tf.function decorator.

My currently solution is set a translating class function and a training class function separately and run get_concrete_function for these two functions separately when saving them.

Is there a way to integrate these two functions into the model’s call method, and use a python argument True/False to control it after I save the model and reload it?

More generally, there is a reserved argument named “training” in Karas Layers and Karas Model, which is a python argument. if I wanna decorate the call method of a Karas Model, how should I set the input signature for this “training” argument?

**Comment by foxik** [comment]
Not a developer here; if I understand it correctly, the Python (non-tf.Tensor) arguments can be used only when the `input_signature` is not explicitly specified. Quoting from https://www.tensorflow.org/api_docs/python/tf/function :
> `input_signature` A possibly nested sequence of tf.TensorSpec objects specifying the shapes and dtypes of the Tensors that will be supplied to this function. If None, a separate function is instantiated for each inferred input signature.  If input_signature is specified, every input to func must be a Tensor, and func cannot accept **kwargs.

In other words, if `input_signature` is given, there is always just a single concrete function.

So the question is whether you must specify the `input_signature` -- in the past, it was needed if you wanted to support dynamic shapes with a single concrete function, but that can be alleviated today using the `experimental_relax_shapes`. Also, the `experimental_follow_type_hints` can play a part of the role of `input_signatures` -- it specifies that some arguments are always Tensors, but unfortunately not their size.

**Comment by LionQYW** [comment]
Emmmm, what about the training argument?

**Comment by gowthamkpr** [comment]
@LionQYW If you want to pass in the training argument, you can do something like this as mentioned in the solution [here](https://stackoverflow.com/questions/66062501/tensorflow-tf-function-conditionals)
